sympy/physics/vector/vector.py

```diff
@@ -684,18 +684,52 @@ def subs(self, *args, **kwargs):
         return Vector(d)
 
     def magnitude(self):
-        """Returns the magnitude (Euclidean norm) of self.
+        """Returns the magnitude (Euclidean 2-norm) of the vector.
+
+        Examples
+        ========
+
+        >>> from sympy import symbols, sqrt
+        >>> from sympy.physics.vector import ReferenceFrame
+        >>> A = ReferenceFrame('A')
+        >>> (1*A.x + 2*A.y + 3*A.z).magnitude()
+        sqrt(14)
+        >>> a, b, c = symbols('a, b, c')
+        >>> (a*A.x + b*A.y + c*A.z).magnitude()
+        sqrt(Abs(a)**2 + Abs(b)**2 + Abs(c)**2)
+
+        SymPy assumes all symbols are complex valued, so you'll need to apply
+        the ``real`` assumption to get the expected result when using
+        magnitude:
+
+        >>> a, b, c = symbols('a, b, c', real=True)
+        >>> (a*A.x + b*A.y + c*A.z).magnitude()
+        sqrt(a**2 + b**2 + c**2)
+
+        This is a vector of length ``a`` but ``a`` can be any real number,
+        positive, negative, or zero. So the magnitude retains the absolute
+        value:
+
+        >>> v = a*sqrt(2)/2*A.x + a*sqrt(2)/2*A.y
+        >>> v.magnitude()
+        Abs(a)
 
         Warnings
         ========
 
-        Python ignores the leading negative sign so that might
-        give wrong results.
-        ``-A.x.magnitude()`` would be treated as ``-(A.x.magnitude())``,
-        instead of ``(-A.x).magnitude()``.
+        Python ignores the leading negative sign so that might give wrong
+        results. ``-A.x.magnitude()`` would be treated as
+        ``-(A.x.magnitude())``, instead of ``(-A.x).magnitude()``.
 
         """
-        return sqrt(self & self)
+        if self.args:
+            col_vec = self.to_matrix(self.args[0][1])
+            summ = S(0)
+            for component in col_vec[:]:
+                summ += abs(component)**2
+            return sqrt(summ)
+        else:  # self is a vector of zero length
+            return S(0)
```

Comment:
I think this magnitude function should assume all the involved symbols are real. It sorta did that before, but missed returning `Abs(measure_number)` for the cases where the sign matters. I could try to special case that scenario, I think that is the only case. Maybe this is also a good spot for the "new" assumptions, where I can do something like:
```python
with all_symbols_in_expressions as real:
    calculate_magnitude()
```
It's a bit annoying to have to apply `real=True` only if you have .magnitude() or .normalize() in your script. This will also be a backwards incompatible output that may cause some downstream confusion.

Reply:
Why do you want them to be assumed as real? If it is to get something to evaluate to a particular form then why not just return that form directly?

Reply:
I want all symbols and functions in the measure numbers of the vector to be assumed real.

> If it is to get something to evaluate to a particular form then why not just return that form directly?

There may be a way to do this, but I am not yet sure of the mathematics to guaranteed the correct solution is returned. If all symbols/functions are are real, then `Matrix.norm()` returns the correct solution.

Reply:
This is what I want:

```python
with vector.free_symbols as real) and (find_dynamisymbols(vector) as real:
    vector.to_matrix(frame).norm()
```

Reply:
What is the actual expression you get if the symbols are not assumed to be real and how does it look if they are assumed to be real?

Reply:
It can really be anything. The vector measure numbers can be arbitrary scalar expressions.

Reply:
But can you give an example? I don't understand what effect you are trying to achieve by asking to assume that the symbols are real.

Reply:
This is the example in the associated issue:

```python
In [1]: import sympy.physics.mechanics as me

In [2]: theta = me.dynamicsymbols('theta')
   ...: 
   ...: A = me.ReferenceFrame('A')
   ...: B = me.ReferenceFrame('B')
   ...: B.orient_axis(A, theta, A.x + A.y)

In [3]: B.ang_vel_in(A).magnitude()
Out[3]: sqrt(Derivative(theta(t), t)**2)
```

It should output `Abs(Derivative(theta(t), t)`.

These are example expressions for the vector components:

```python
In [3]: B.ang_vel_in(A).to_matrix(B)
Out[3]: 
Matrix([
[sqrt(2)*(1/2 - cos(theta(t))/2)*Derivative(theta(t), t)/2 + sqrt(2)*(cos(theta(t))/2 + 1/2)*Derivative(theta(t), t)/2],
[sqrt(2)*(1/2 - cos(theta(t))/2)*Derivative(theta(t), t)/2 + sqrt(2)*(cos(theta(t))/2 + 1/2)*Derivative(theta(t), t)/2],
[                                                                                                                    0]])
```

Reply:
Here is another:
```python
In [4]: B.orient_axis(A, theta, A.x + A.y + A.z)

In [5]: B.ang_vel_in(A).to_matrix(B)
Out[5]: 
Matrix([
[sqrt(3)*(2*cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3 + sqrt(3)*(-sqrt(3)*sin(theta(t))/3 - cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3 + sqrt(3)*(sqrt(3)*sin(theta(t))/3 - cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3],
[sqrt(3)*(2*cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3 + sqrt(3)*(-sqrt(3)*sin(theta(t))/3 - cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3 + sqrt(3)*(sqrt(3)*sin(theta(t))/3 - cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3],
[sqrt(3)*(2*cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3 + sqrt(3)*(-sqrt(3)*sin(theta(t))/3 - cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3 + sqrt(3)*(sqrt(3)*sin(theta(t))/3 - cos(theta(t))/3 + 1/3)*Derivative(theta(t), t)/3]])
```

Reply:
So the 2-norm is `sqrt(Abs(x)**2 + Abs(y)**2 + Abs(z)**2)` for complex valued x, y, and z (at least that is what `Matrix.norm()` does.

But we have the naive implementation now: `sqrt(x**2 + y**2 + z**2)` which works for most cases. But for some values of `x, y, z` this is not sufficient to get the correct answer (see two examples above). But if you know x, y, z are real valued expressions, then I get the correct 2 norm.

So I'm not sure what the 2 norm calculation should be such that real is assumed.

Reply:
Basically any vector that looks like `v = some_arbitary_expression*(A.x + A.y + A.z)` should return:

```
>>> v.magnitude()
Abs(some_arbitrary_expression)
```
So I guess the special case I'm looking for is if the vector has a simple scalar multiplicative. I should be able to check for that somehow.

Reply:
I would calculate this as the squared norm. Then if you get a power you can "unsquare" it but otherwise you just return the square root e.g.:
```python
In [48]: mysqrt = lambda obj: Abs(obj.base) if obj.is_Pow and obj.exp == 2 else obj

In [49]: v = B.ang_vel_in(A)

In [50]: v.dot(v)
Out[50]: 
          2
⎛d       ⎞ 
⎜──(θ(t))⎟ 
⎝dt      ⎠ 

In [51]: mysqrt(_)
Out[51]: 
│d       │
│──(θ(t))│
│dt      │
```

Reply:
Actually maybe all that's needed is just to apply the abs before taking the norm e.g.:
```python
In [91]: B.ang_vel_in(A).to_matrix(A).applyfunc(Abs).norm()
Out[91]: 
│d       │
│──(θ(t))│
│dt      │
```

Reply:
But that will give the wrong answer for simple vectors:

```
In [7]: a, b, c = sm.symbols('a, b, c')

In [8]: v = a*A.x + b*A.y + c*A.z

In [9]: v.to_matrix(A).applyfunc(Abs).norm()
Out[9]: sqrt(Abs(a)**2 + Abs(b)**2 + Abs(c)**2)
```

Reply:
We want:
```
In [10]: a, b, c = sm.symbols('a, b, c', real=True)

In [11]: v = a*A.x + b*A.y + c*A.z

In [12]: v.to_matrix(A).applyfunc(Abs).norm()
Out[12]: sqrt(a**2 + b**2 + c**2)

```

Reply:
Okay, I see now. The norm over a vector of complex numbers would usually be defined as `sqrt(dot(v, conjugate(v))` which is the same as `sqrt(abs(v1)**2+abs(v2)**2+...)`. That definition with abs is also valid for real numbers but with real numbers we have other identities that can simplify this like
```
sqrt(abs(x)**2)  ->  abs(x)
sqrt(x**2)  ->  abs(x)
abs(x)**2  ->  x**2
```
Those identities could all be applied manually. I think I would be more inclined though to just compute the norm and then see if ends up having something squared that can be turned into an abs.

Reply:
Ok, I'll think about it and see if I can implement a custom processor to apply these things without having to apply the `real=True` assumption.